architecture/2024-10-user-defined-types-authoring-experience-feature-spec.md

```diff
@@ -0,0 +1,230 @@
+                connectionString:
+                    type: 'string'
+                    description: 'Connection string to the messaging service'
+                logAnalytics:
+                    type: 'string'
+                    description: 'Usage type of the messaging service'  
+                    default: 'off'        
+```
+
+Deb creates the updated resource type schema via the CLI
+
+```bash
+rad resource-type create -f plaid.yaml
+Resource type `plaid` created successfully for version `2024-11-01`
+```
+
+Question 
+
+1. We leave it up to the user to decide whether to use the new optional property in their application definitions and Recipes. Should we provide any guidance?
+2. If Deb tries to add an optional property within the same version or without a default value, what should be the behavior ? Should we consider that any change warrants a api version upgrade like Azure/ ARM does?
+
+#### Sub-scenario 2: Deb adds a required property and updates the resource type Plaid.
+Deb wants to add a required property `messageForwarding`. This is considered to be a breaking change as it would require the Infrastructure operator teams and developers to update their Recipes and application definitions to use the new property in the resource type Plaid. Deb should be provided with an error message when a breaking change is detected in the schema definition.
+
+```yaml
+namespace: 'Mycompany.Messaging'
+resourceTypes:
+    'Mycompany.Messaging/plaidResource':
+        name: 'Name of the resource'
+        version: '2024-11-01'
+            properties: 
+                required: ['host', 'port', 'connectionString']
+                queueName:
+                    type: 'string'
+                    description: 'Name of the queue'   
+                host:
+                    type: 'string'
+                    description: 'Hostname of the messaging service'        
+                port:
+                    type: 'string'
+                    description: 'Port'
+                connectionString:
+                    type: 'string'
+                    description: 'Connection string to the messaging service'
+                messageForwarding:
+                    type: 'string'
+                    description: 'Usage type of the messaging service'          
+```
+
+Deb creates the updated resource type schema via the CLI
+
+```bash
+rad resource-type create -f plaid.yaml
+Error: Unable to create the resource-type `plaid` for version `2024-10-01`. The schema definitions has a breaking change with the required property `instance_type`. Please delete older versions of the resource type and try again.
+```
+
+Questions
+1. For MVP, can we just flag breaking changes and provide documentation on how to handle breaking changes ? 
+2. What are the breaking changes that we need to consider ? 
+    - Adding a required property
+    - Removing a required property
+    - Changing the type of a property
+    - Changing the name of a property
```

Comment:
Trying to precisely define each type of breaking change is a task for writing the documentation. We should define it in-terms of assignability.

- Valid data from an older API version can be *assigned* to a new version and still be valid. 
- Valid data from an new API version can be *assigned* to an old version and still be valid. 

Fields may be dropped during the conversion process while moving from new -> old.